crates/saya/core/src/prover/scheduler.rs

```diff
@@ -0,0 +1,167 @@
+use anyhow::bail;
+use cairo_proof_parser::output::{extract_output, ExtractOutputResult};
+use futures::future::BoxFuture;
+use futures::FutureExt;
+use katana_primitives::state::StateUpdates;
+use katana_primitives::FieldElement;
+use tokio::sync::oneshot;
+use tracing::{info, trace};
+
+use super::{prove_diff, ProgramInput, ProverIdentifier};
+use crate::prover::extract::program_input_from_program_output;
+use crate::prover::ProveProgram;
+use crate::LOG_TARGET;
+
+type Proof = String;
+
+pub struct Scheduler {
+    root_task: BoxFuture<'static, anyhow::Result<(Proof, ProgramInput)>>,
+    free_differs: Vec<oneshot::Sender<ProgramInput>>,
+}
+
+impl Scheduler {
+    pub fn new(capacity: usize, world: FieldElement, prover: ProverIdentifier) -> Self {
+        let (senders, receivers): (Vec<_>, Vec<_>) =
+            (0..capacity).map(|_| oneshot::channel::<ProgramInput>()).unzip();
+
+        let root_task = prove_recursively(receivers, world, prover);
+
+        Scheduler { root_task, free_differs: senders }
+    }
+
+    pub fn is_full(&self) -> bool {
+        self.free_differs.is_empty()
+    }
+
+    pub fn push_diff(&mut self, input: ProgramInput) -> anyhow::Result<()> {
+        if self.is_full() {
+            bail!("Scheduler is full");
+        }
+
+        let sender = self.free_differs.remove(0);
+        if sender.send(input).is_err() {
+            bail!("Failed to send input to differ");
+        }
+        Ok(())
+    }
+
+    pub async fn proved(self) -> anyhow::Result<(Proof, ProgramInput)> {
+        self.root_task.await
+    }
+
+    pub async fn merge(
+        inputs: Vec<ProgramInput>,
+        world: FieldElement,
+        prover: ProverIdentifier,
+    ) -> anyhow::Result<(Proof, ProgramInput)> {
+        let mut scheduler = Scheduler::new(inputs.len(), world, prover);
+        trace!(target: LOG_TARGET, "pushing inputs to scheduler");
+        for input in inputs {
+            scheduler.push_diff(input)?;
+        }
+        info!(target: LOG_TARGET, "inputs pushed to scheduler");
+        scheduler.proved().await
+    }
+}
+
+#[derive(serde::Serialize, serde::Deserialize)]
+struct CombinedInputs {
+    earlier: ProgramInput,
+    later: ProgramInput,
+}
+
+async fn combine_proofs(
+    first: Proof,
+    second: Proof,
+    prover: ProverIdentifier,
+    state_updates1: StateUpdates,
+    state_updates2: StateUpdates,
+    world: FieldElement,
+) -> anyhow::Result<Proof> {
+    let ExtractOutputResult { program_output: program_output1, program_output_hash: _ } =
+        extract_output(&first)?;
+    let ExtractOutputResult { program_output: program_output2, program_output_hash: _ } =
+        extract_output(&second)?;
+
+    let earlier_input =
+        program_input_from_program_output(program_output1, state_updates1, world).unwrap();
+    let later_input =
+        program_input_from_program_output(program_output2, state_updates2, world).unwrap();
+
+    trace!(target: LOG_TARGET, "Merging proofs");
```

Comment:
Do we have a program hash or fact or something to identify who's being merged to add this to the log?

Reply:
Number of inputs and world should help.